capella2polarion/connectors/polarion_worker.py

```diff
@@ -425,36 +461,104 @@ def compare_and_update_work_items(
+    def update_documents(
+        self,
+        document_datas: list[data_models.DocumentData],
+        document_project: str | None = None,
+    ):
         """Update existing documents."""
-        self.client.project_client.documents.update(documents)
+        client = self._get_client(document_project)
+        documents, headings = self._process_document_datas(
+            client, document_datas
+        )
+
+        client.work_items.update(headings)
+        client.documents.update(documents)
+
+    def _process_document_datas(
+        self,
+        client: polarion_api.ProjectClient,
+        document_datas: list[data_models.DocumentData],
+    ):
+        documents = []
+        headings = []
+        for document_data in document_datas:
+            headings += document_data.headings
+            documents.append(document_data.document)
+            if document_data.text_work_item_provider.new_text_work_items:
+                self._create_and_update_text_work_items(
+                    document_data.text_work_item_provider.new_text_work_items,
+                    client,
+                )
+                document_data.text_work_item_provider.insert_text_work_items(
+                    document_data.document,
+                )
+        return documents, headings
 
     def get_document(
-        self, space: str, name: str
+        self, space: str, name: str, document_project: str | None = None
     ) -> polarion_api.Document | None:
         """Get a document from polarion and return None if not found."""
+        client = self._get_client(document_project)
         try:
-            return self.client.project_client.documents.get(
+            return client.documents.get(
                 space, name, fields={"documents": "@all"}
             )
         except polarion_api.PolarionApiBaseException as e:
             if e.args[0] == 404:
                 return None
             raise e
 
-    def update_work_items(self, work_items: list[polarion_api.WorkItem]):
-        """Update the given workitems without any additional checks."""
-        self.client.project_client.work_items.update(work_items)
-
     def load_polarion_documents(
-        self, document_paths: t.Iterable[tuple[str, str]]
-    ) -> dict[tuple[str, str], polarion_api.Document | None]:
-        """Load the given document references from Polarion."""
+        self,
+        document_infos: t.Iterable[data_models.DocumentInfo],
+    ) -> dict[
+        tuple[str | None, str, str],
+        tuple[polarion_api.Document | None, list[polarion_api.WorkItem]],
+    ]:
+        """Load the documents referenced and text work items from Polarion."""
         return {
-            (space, name): self.get_document(space, name)
-            for space, name in document_paths
+            (di.project_id, di.module_folder, di.module_name): (
+                self.get_document(
+                    di.module_folder, di.module_name, di.project_id
+                ),
```

Comment:
`get_document` could be `None` for whatever reason and it would land in the dictionary. A key `None` is always dangerous, since it tends to be overwritten often, losing the previous values. I'd like to see `None` be gone and this can be done easily while also improving the readability of this return statement.

Reply:
I'd do it like this:
```py
        documents_with_its_work_items = {}
        for di in document_infos:
            document = self.get_document(
                di.module_folder, di.module_name, di.project_id
            )
            work_items = self._get_client(di.project_id).work_items.get_all(
                WORK_ITEMS_IN_PROJECT_QUERY.format(
                    project=di.project_id or self.polarion_params.project_id,
                    doc_folder=di.module_folder,
                    doc_name=di.module_name,
                    wi_type=di.text_work_item_type,
                ),
                fields={"workitems": f"id,{di.text_work_item_id_field}"},
            )
            documents_with_its_work_items[
                (di.project_id, di.module_folder, di.module_name)
            ] = (document, work_items)

        return documents_with_its_work_items
```

Reply:
Sorry, I don't get it. Can you explain this in detail?